rotkehlchen/chain/evm/decoding/zerox/decoder.py

```diff
@@ -65,6 +67,43 @@ def _update_send_and_receive_events(
             receive_event.event_type = HistoryEventType.TRADE
             receive_event.event_subtype = HistoryEventSubType.RECEIVE
             receive_event.notes = f'Receive {receive_event.balance.amount} {receive_event.asset.symbol_or_name()} as the result of a swap via the 0x protocol'  # noqa: E501
+        if fee_event is not None:
+            fee_event.counterparty = CPT_ZEROX
+            fee_event.address = self.router_address
+            fee_event.event_type = HistoryEventType.SPEND
+            fee_event.event_subtype = HistoryEventSubType.FEE
+            fee_event.notes = f'Spend {fee_event.balance.amount} {fee_event.asset.symbol_or_name()} as a 0x protocol fee'  # noqa: E501
+
+    def _merge_similar_events(
+            self,
+            decoded_events: 'list[EvmEvent]',
+            send_events: 'list[EvmEvent]',
+            receive_events: 'list[EvmEvent]',
+            fee_event: 'EvmEvent | None' = None,
+    ) -> None:
+        """Sum the balances of the events, updates them with the 0x values, replace them with the
+        Merged Events in decoded_events list, and maybe shuffle them with proper order."""
+        summed_send_event = send_events[0] if len(send_events) > 0 else None
+        summed_receive_event = receive_events[0] if len(receive_events) > 0 else None
+
+        if summed_send_event is not None:
+            for event in send_events[1:]:
+                summed_send_event.balance += event.balance
+                decoded_events.remove(event)
+        if summed_receive_event is not None:
+            for event in receive_events[1:]:
+                summed_receive_event.balance += event.balance
+                decoded_events.remove(event)
```

Comment:
removing from a list in python irc has `O(n)` time complexity. Instead of this do a set 

```python
events_to_remove = {}
```

and add to that set the identifier of the events. Then you can do

```python
decoded_events = [event for event in decoded_events if event.identifier not in events_to_remove]
```